presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java

```diff
@@ -689,31 +687,103 @@ protected ExtractionResult visitNullLiteral(NullLiteral node, Boolean complement
         Object left = ExpressionInterpreter.expressionOptimizer(comparison.getLeft(), metadata, session, expressionTypes).optimize(NoOpSymbolResolver.INSTANCE);
         Object right = ExpressionInterpreter.expressionOptimizer(comparison.getRight(), metadata, session, expressionTypes).optimize(NoOpSymbolResolver.INSTANCE);
 
-        if (left instanceof SymbolReference && !(right instanceof Expression)) {
-            return Optional.of(new NormalizedSimpleComparison((SymbolReference) left, comparison.getType(), new NullableValue(expressionTypes.get(comparison.getRight()), right)));
+        Type leftType = expressionTypes.get(comparison.getLeft());
+        Type rightType = expressionTypes.get(comparison.getRight());
+        Preconditions.checkArgument(leftType.equals(rightType), "left and right type do not match in comparison expression (%s)", comparison);
+
+        if (left instanceof Expression && right instanceof Expression) {
+            return Optional.empty();
+        }
+
+        if (!(left instanceof Expression) && !(right instanceof Expression)) {
+            return Optional.empty();
+        }
+
+        Expression symbolExpression;
+        ComparisonExpressionType comparisonType;
+        NullableValue value;
+
+        if (left instanceof Expression) {
+            symbolExpression = comparison.getLeft();
+            comparisonType = comparison.getType();
+            value = new NullableValue(rightType, right);
+        }
+        else {
+            symbolExpression = comparison.getRight();
+            comparisonType = comparison.getType().flip();
+            value = new NullableValue(leftType, left);
+        }
+
+        if (!checkOnlyImplicitCastsAroundSymbol(metadata, expressionTypes, symbolExpression)) {
```

Comment:
instead of checking the implicit casts here and having that information be retained in the NormalizedSimpleComparison by convention, it seems cleaner just to return the arbitrary Expression in NormalizedSimpleComparison, and have the calling code do the implicit coercion check as it is unpeeling the casts.